### PR TITLE
feat: with-timing

### PR DESCRIPTION
**Pull request by thiskevinwang**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-BRANCH_NAME-hashicorp.vercel.app/PATH_TO_VIEW) 🔎
- [Asana task](https://app.asana.com/0/1100423001970639/1203442542938543/f) 🎟️

## 🗒️ What

This adds a `with-timing` method to wrap function calls, and apply a label for console debugging.

This also applies the method to a small handful of places across the codebase.

The logging is toggled via env var: `HC_DEBUG_TIMINGS="1"`

<details>
  <summary><b>Console Output on:</b> <code>http://localhost:3000/consul/docs</code></summary>

```console
wait  - compiling /api/auth/[...nextauth] (client and server)...
event - compiled successfully in 75 ms (1093 modules)
[next-auth][warn][NEXTAUTH_URL] 
https://next-auth.js.org/warnings#nextauth_url
[middleware] /consul/docs: 0.074ms
[[productSlug]/docs/index::getStaticProps] (consul)
  [docs-view/server::loadStaticProps]
    [getTutorialMap] (http://localhost:3000/api/tutorials-map)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100): 57.059ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=2b5a0644-8b93-4dba-ac5c-7843de4b1289)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=2b5a0644-8b93-4dba-ac5c-7843de4b1289): 36.927ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=0af4d72f-51dd-4dc5-85e4-67c68030ce8d)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=0af4d72f-51dd-4dc5-85e4-67c68030ce8d): 42.007ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=8bd94156-cf0a-4f88-b5f5-d2ffa57f6402)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=8bd94156-cf0a-4f88-b5f5-d2ffa57f6402): 43.387ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=e48563d4-00d1-4c0b-a5da-13e276ab119c)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=e48563d4-00d1-4c0b-a5da-13e276ab119c): 37.064ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=75bb9296-0364-4ec4-8fbd-3930c9e947d6)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=75bb9296-0364-4ec4-8fbd-3930c9e947d6): 42.021ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=216b3270-fc89-4ad5-9053-603be0df67e8)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=216b3270-fc89-4ad5-9053-603be0df67e8): 39.374ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=95a7e36f-44ac-4e39-9fb3-c25db26f4af9)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=95a7e36f-44ac-4e39-9fb3-c25db26f4af9): 49.598ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=6d51c189-d38e-493f-b373-7e2dfe6bdbc5)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=6d51c189-d38e-493f-b373-7e2dfe6bdbc5): 40.674ms
    [getTutorialMap] (http://localhost:3000/api/tutorials-map): 542.559ms
  [docs-view/server::loadStaticProps]: 631.926ms
  [docs-view/server::prepareNavDataForClient]
    [getTutorialMap] (http://localhost:3000/api/tutorials-map)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100): 42.584ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=2b5a0644-8b93-4dba-ac5c-7843de4b1289)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=2b5a0644-8b93-4dba-ac5c-7843de4b1289): 41.721ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=0af4d72f-51dd-4dc5-85e4-67c68030ce8d)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=0af4d72f-51dd-4dc5-85e4-67c68030ce8d): 50.425ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=8bd94156-cf0a-4f88-b5f5-d2ffa57f6402)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=8bd94156-cf0a-4f88-b5f5-d2ffa57f6402): 46.331ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=e48563d4-00d1-4c0b-a5da-13e276ab119c)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=e48563d4-00d1-4c0b-a5da-13e276ab119c): 48.954ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=75bb9296-0364-4ec4-8fbd-3930c9e947d6)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=75bb9296-0364-4ec4-8fbd-3930c9e947d6): 47.509ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=216b3270-fc89-4ad5-9053-603be0df67e8)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=216b3270-fc89-4ad5-9053-603be0df67e8): 36.666ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=95a7e36f-44ac-4e39-9fb3-c25db26f4af9)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=95a7e36f-44ac-4e39-9fb3-c25db26f4af9): 39.325ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=6d51c189-d38e-493f-b373-7e2dfe6bdbc5)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=6d51c189-d38e-493f-b373-7e2dfe6bdbc5): 34.511ms
    [getTutorialMap] (http://localhost:3000/api/tutorials-map): 479.168ms
  [docs-view/server::prepareNavDataForClient]: 482.362ms
[[productSlug]/docs/index::getStaticProps] (consul): 1.116s
```

</details>

### CODE REVIEW

File: src/lib/with-timing.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/**
+ * A utility method to wrap a function, and log execution time.
+ */
+export async function withTiming<T>(
+	label: string,
+	fn: () => Promise<T> | T
+): Promise<T> {
```
Comment (BRKalow) [inline src/lib/with-timing.ts:7]: 💭 I like the grouping, but it also introduces extra noise for "leaf" timings. E.g. when I'm wrapping a fetch call in `withTiming` I don't necessarily want to see the two log lines that coming with using `console.group()`. What do you think about adding an option to disable the grouping behavior?

Comment (thiskevinwang) [reply src/lib/with-timing.ts:7]: yea, I was grappling with this a little bit...

And are you referring to this doubling-up?

CleanShot 2022-11-29 at 17 16 05@2x

Comment (BRKalow) [reply src/lib/with-timing.ts:7]: Yeah, exactly 👍

Comment (thiskevinwang) [reply src/lib/with-timing.ts]: @BRKalow -> https://github.com/hashicorp/dev-portal/pull/1395/commits/bc5ed7ccf8fbea329c56976a0d0a5bff97aaf91c

```console
[middleware] /consul/docs: 0.242ms
[[productSlug]/docs/index::getStaticProps] (consul)
  [docs-view/server::loadStaticProps]
    [getTutorialMap] (http://localhost:3000/api/tutorials-map)
      wait  - compiling /api/tutorials-map (client and server)...
      event - compiled successfully in 53 ms (1077 modules)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100): 118.432ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=2b5a0644-8b93-4dba-ac5c-7843de4b1289): 39.375ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=0af4d72f-51dd-4dc5-85e4-67c68030ce8d): 46.31ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=8bd94156-cf0a-4f88-b5f5-d2ffa57f6402): 39.629ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=e48563d4-00d1-4c0b-a5da-13e276ab119c): 38.701ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=75bb9296-0364-4ec4-8fbd-3930c9e947d6): 49.063ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=216b3270-fc89-4ad5-9053-603be0df67e8): 43.516ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=95a7e36f-44ac-4e39-9fb3-c25db26f4af9): 58.987ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=6d51c189-d38e-493f-b373-7e2dfe6bdbc5): 50.253ms
    [getTutorialMap] (http://localhost:3000/api/tutorials-map): 1.142s
  [docs-view/server::loadStaticProps]: 1.234s
  [docs-view/server::prepareNavDataForClient]
    [getTutorialMap] (http://localhost:3000/api/tutorials-map)
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100): 48.933ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=2b5a0644-8b93-4dba-ac5c-7843de4b1289): 37.374ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=0af4d72f-51dd-4dc5-85e4-67c68030ce8d): 33.722ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=8bd94156-cf0a-4f88-b5f5-d2ffa57f6402): 38.009ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=e48563d4-00d1-4c0b-a5da-13e276ab119c): 44.829ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=75bb9296-0364-4ec4-8fbd-3930c9e947d6): 45.542ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=216b3270-fc89-4ad5-9053-603be0df67e8): 50.461ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=95a7e36f-44ac-4e39-9fb3-c25db26f4af9): 39.352ms
      [lib/learn-client] (GET https://ycmsw3yg4a.us-east-1.awsapprunner.com/tutorials/?limit=100&after=6d51c189-d38e-493f-b373-7e2dfe6bdbc5): 34.98ms
    [getTutorialMap] (http://localhost:3000/api/tutorials-map): 860.337ms
  [docs-view/server::prepareNavDataForClient]: 863.274ms
[[productSlug]/docs/index::getStaticProps] (consul): 2.099s
```

**Review by BRKalow — APPROVED** [review]
👏

**Pull request merged**